### PR TITLE
Update scoring API to allow callable object / function with signature scorer(estimator, X, y)

### PR DESCRIPTION
**Pull request by weixuanfu**
## What does this PR do?

1. Update scoring API to allow callable object / function with `signature scorer(estimator, X, y)`

2. Raise `DeprecationWarning` if user use a metric function instead of scorer function. And metric function will be deprecated in version 0.11

3. Add unit tests for new API

4. Update docs and API .

## Where should the reviewer start?

`_setup_scoring_function` in base.py

## How should this PR be tested?

```Python
from tpot import TPOTClassifier
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics.scorer import make_scorer

digits = load_digits()
X_train, X_test, y_train, y_test = train_test_split(digits.data, digits.target,
                                                    train_size=0.75, test_size=0.25)
# make a custom metric function
def my_custom_accuracy(y_true, y_pred):
    return float(sum(y_pred == y_true)) / len(y_true)
# make a custom a scorer from the custom metric function
my_custom_scorer = make_scorer(my_custom_accuracy, greater_is_better=True)

tpot = TPOTClassifier(generations=5, population_size=20, verbosity=2,
                      scoring=my_custom_scorer)
tpot.fit(X_train, y_train)
print(tpot.score(X_test, y_test))
tpot.export('tpot_mnist_pipeline.py')
```

## What are the relevant issues?

#579 

## Questions:

- Do the docs need to be updated? No
- Does this PR add new (Python) dependencies? No

### CODE REVIEW

**Comment by weixuanfu** [comment]
Update: update 2 time-consuming unit tests for passing CI in python 2.7 (issue #588 )

**Comment by rhiever** [comment]
There's merge conflicts due to the recent PR merges.

File: docs_sources/api.md [hunk 1]
```diff
@@ -75,9 +75,12 @@ Function used to evaluate the quality of a given pipeline for the classification
 'accuracy', 'adjusted_rand_score', 'average_precision', 'balanced_accuracy', 'f1', 'f1_macro', 'f1_micro', 'f1_samples', 'f1_weighted', 'neg_log_loss','precision',
 'precision_macro', 'precision_micro', 'precision_samples', 'precision_weighted', 'recall', 'recall_macro', 'recall_micro', 'recall_samples', 'recall_weighted', 'roc_auc'
 <br /><br/>
-If you would like to use a custom scoring function, you can pass a callable function to this parameter with the signature <em>scorer(y_true, y_pred)</em>. See the section on <a href="../using/#scoring-functions">scoring functions</a> for more details.
-<br /><br />
-TPOT assumes that any function with "error" or "loss" in the function name is meant to be minimized, whereas any other functions will be maximized.
+If you would like to use a custom scorer, you can pass the callable object/function with signature <em>scorer(estimator, X, y)</em>.
```
Comment (rhiever) [inline docs_sources/api.md:78]: It's just a callable function, not an object.

File: docs_sources/using.md [hunk 2]
```diff
@@ -350,28 +350,34 @@ TPOT makes use of `sklearn.model_selection.cross_val_score` for evaluating pipel
...
-
-tpot = TPOTClassifier(generations=5, population_size=20, verbosity=2,
-                      scoring=my_custom_accuracy)
-tpot.fit(X_train, y_train)
-print(tpot.score(X_test, y_test))
-tpot.export('tpot_mnist_pipeline.py')
-```
-
-* **my_module.scorer_name**: you can also use your manual  `scorer(y_true, y_pred)` function through the command line, just add an argument `-scoring my_module.scorer` and TPOT will import your module and take the function from there. TPOT will also include current workdir when importing the module, so you can just put it in the same folder where you are going to run.
+2. you can pass the callable object/function with signature `scorer(estimator, X, y)`, where `estimator` is trained estimator to use for scoring, `X` are features that will be fed to `estimator.predict` and `y` are target values for `X`. To do this, you should implement your own function. See the example below for further explanation.
```
Comment (rhiever) [inline docs_sources/using.md]: Capitalize 'You' at the beginning.

Change "fed" --> "passed"

File: docs_sources/using.md [hunk 3]
```diff
@@ -350,28 +350,34 @@ TPOT makes use of `sklearn.model_selection.cross_val_score` for evaluating pipel
...
+  from sklearn.metrics.scorer import make_scorer
+
+  digits = load_digits()
+  X_train, X_test, y_train, y_test = train_test_split(digits.data, digits.target,
+                                                      train_size=0.75, test_size=0.25)
+  # make a custom metric function
+  def my_custom_accuracy(y_true, y_pred):
+      return float(sum(y_pred == y_true)) / len(y_true)
+  # make a custom a scorer from the custom metric function
+  my_custom_scorer = make_scorer(my_custom_accuracy, greater_is_better=True)
```
Comment (rhiever) [inline docs_sources/using.md]: Maybe add a note in the comments here that `greater_is_better=False` would mean that the scoring function should be minimized.

File: docs_sources/using.md [hunk 4]
```diff
@@ -350,28 +350,34 @@ TPOT makes use of `sklearn.model_selection.cross_val_score` for evaluating pipel
...
+
+  tpot = TPOTClassifier(generations=5, population_size=20, verbosity=2,
+                        scoring=my_custom_scorer)
+  tpot.fit(X_train, y_train)
+  print(tpot.score(X_test, y_test))
+  tpot.export('tpot_mnist_pipeline.py')
+  ```
+
+3. You can pass a metric function with the signature `score_func(y_true, y_pred)` (e.g. `my_custom_accuracy` in the example above), where `y_true` are the true target values and `y_pred` are the predicted target values from an estimator. To do this, you should implement your own function. See the example below for further explanation. TPOT assumes that any function with "error" or "loss" in the function name is meant to be minimized (`greater_is_better=False` in [`make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html)), whereas any other functions will be maximized. This scoring type was deprecated in version 0.9.1 and will be removed in version 0.11.
```
Comment (rhiever) [inline docs_sources/using.md]: The example referenced is above, not below.

File: docs_sources/using.md [hunk 5]
```diff
@@ -350,28 +350,34 @@ TPOT makes use of `sklearn.model_selection.cross_val_score` for evaluating pipel
...
+
+
+* **my_module.scorer_name**: you can also use your manual  `score_func(y_true, y_pred)` or `scorer(estimator, X, y)` function through the command line, just add an argument `-scoring my_module.scorer` and TPOT will import your module and take the function from there. TPOT will also include current workdir when importing the module, so you can just put it in the same folder where you are going to run.
```
Comment (rhiever) [inline docs_sources/using.md]: "TPOT will also include current workdir when importing the module, so you can just put it in the same folder where you are going to run." 

-->

TPOT will include your current working directory when importing the module, so you can place it in the same directory where you are going to run TPOT.

File: docs_sources/using.md (hunk 4 above)
Comment (weixuanfu) [reply docs_sources/using.md]: It should be "above" since there is a example below 2. but above 3.

File: docs_sources/api.md (hunk 1 above)
Comment (weixuanfu) [reply docs_sources/api.md:78]: I copied [this line](https://github.com/scikit-learn/scikit-learn/blob/ef5cb84a/sklearn/model_selection/_validation.py#L241)

File: tpot/base.py [hunk 6]
```diff
@@ -351,6 +336,35 @@ def __init__(self, generations=100, population_size=100, offspring_size=None,
...
+                        'choose a valid scoring function from the TPOT '
+                        'documentation.'.format(scoring)
+                    )
+            elif callable(scoring):
+                if not isinstance(scoring, _BaseScorer):
+                    # If the user passed a custom metric function, store it in the sklearn
+                    # SCORERS dictionary
+                    scoring_name = scoring.__name__
+                    greater_is_better = 'loss' not in scoring_name and 'error' not in scoring_name
+                    SCORERS[scoring_name] = make_scorer(scoring, greater_is_better=greater_is_better)
```
Comment (rhiever) [inline tpot/base.py:357]: Since `self.scoring_function` is now a string (name of one of the default scoring functions) OR a scoring function with the signature `my_scorer(estimator, X, y)`, perhaps we shouldn't store custom scoring functions in the `SCORERS` dictionary any more. We can store the result of this `make_scorer` call directly to `self.scoring_function`.

File: docs_sources/api.md (hunk 1 above)
Comment (rhiever) [reply docs_sources/api.md:78]: OK.

File: docs_sources/using.md (hunk 4 above)
Comment (rhiever) [reply docs_sources/using.md]: Correct. This is for 3.---it says "below".

File: tpot/base.py (hunk 6 above)
Comment (rhiever) [reply tpot/base.py:357]: Actually, we need to consider the broader use of the `SCORERS` dictionary. If someone passes a scoring function with the signature `my_scorer(estimator, X, y)`, I don't think it would work in the TPOT `score` function. Perhaps all custom scorers need to be stored in the `SCORERS` dictionary.

See [here](https://github.com/weixuanfu/tpot/blob/8252c742514f24bbe12e837534c0d21ea8caa09b/tpot/base.py#L843).

We should also have unit tests that cover all of these possible custom scorer input cases.

File: docs_sources/using.md (hunk 4 above)
Comment (weixuanfu) [reply docs_sources/using.md]: Gotcha, I just found it. there is a `above` before that. I corrected that one but missed this one.

File: docs_sources/using.md [hunk 7]
```diff
@@ -350,28 +350,36 @@ TPOT makes use of `sklearn.model_selection.cross_val_score` for evaluating pipel
...
 
-tpot = TPOTClassifier(generations=5, population_size=20, verbosity=2,
-                      scoring=my_custom_accuracy)
-tpot.fit(X_train, y_train)
-print(tpot.score(X_test, y_test))
-tpot.export('tpot_mnist_pipeline.py')
-```
 
-* **my_module.scorer_name**: you can also use your manual  `scorer(y_true, y_pred)` function through the command line, just add an argument `-scoring my_module.scorer` and TPOT will import your module and take the function from there. TPOT will also include current workdir when importing the module, so you can just put it in the same folder where you are going to run.
+* **my_module.scorer_name**: you can also use your manual  `score_func(y_true, y_pred)` or `scorer(estimator, X, y)` function through the command line, just add an argument `-scoring my_module.scorer` and TPOT will import your module and take the function from there. TPOT will include your current working directory when importing the module, so you can place it in the same directory where you are going to run TPOT.
```
Comment (rhiever) [inline docs_sources/using.md]: >you can also use your manual  `score_func(y_true, y_pred)` or `scorer(estimator, X, y)` function through the command line, just add an argument `-scoring my_module.scorer` and TPOT will import your module and take the function from there.

-->

>You can also use a custom  `score_func(y_true, y_pred)` or `scorer(estimator, X, y)` function through the command line by adding the argument `-scoring my_module.scorer` to your command-line call. TPOT will import your module and use the custom scoring function from there.

File: tpot/base.py (hunk 6 above)
Comment (weixuanfu) [reply tpot/base.py:357]: OK, will change it

Comment (weixuanfu) [reply tpot/base.py:357]: fixed in commit below. I think unit tests already cover all these scoring input types

Comment (rhiever) [reply tpot/base.py:357]: Looks the same to me; did the commit get pushed?

File: tpot/base.py [hunk 8]
```diff
@@ -353,6 +338,40 @@ def __init__(self, generations=100, population_size=100, offspring_size=None,
...
+                    warnings.warn('Scoring value {} looks like it is a metric function '
+                                    'rather than a scorer. This scoring type was deprecated '
+                                    'in version 0.9.1 and will be removed in version 0.11. '
+                                    'TPOT uses `make_scorer` to convert this metric function'
+                                    'to a scorer which requires an estimator as its first '
+                                    'parameter'.format(scoring), DeprecationWarning)
+
+                else:
+                    scoring_name = scoring._score_func.__name__
+                    SCORERS[scoring_name] = scoring
```
Comment (weixuanfu) [inline tpot/base.py]: @rhiever Here is the case of using board `SCORERS` dict for a scoring function with the signature `my_scorer(estimator, X, y)`

File: tpot/base.py (hunk 6 above)
Comment (weixuanfu) [reply tpot/base.py:357]: Yep, check the comment below

File: tpot/base.py (hunk 8 above)
Comment (rhiever) [reply tpot/base.py]: OK, I see it now. Must've taken a minute to update.
